hawkmoth/__init__.py

```diff
@@ -62,15 +63,54 @@ def __display_parser_diagnostics(self, errors):
                 self.logger.log(self._log_lvl[severity], toprint,
                                 location=(env.docname, self.lineno))
 
-    def __get_transform(self):
+    def __get_compat_transform(self):
         env = self.state.document.settings.env
 
         compat = self.options.get('compat', env.config.cautodoc_compat)
         if compat is None:
             return None
 
+        fmt = 'cautodoc_compat and compat options are deprecated, please use cautodoc_transformations and transform options instead.'
+        self.logger.warning(fmt, location=(env.docname, self.lineno))
+
         return lambda comment: doccompat.convert(comment, transform=compat)
 
+    def __get_transform(self):
+        env = self.state.document.settings.env
+
+        # Special case values for transform.
+        OPT_NONE = 'none'
+        OPT_DEFAULT = 'default'
+
+        # Handle deprecated compat. To be removed.
+        transform = self.__get_compat_transform()
+        if transform is not None:
+            return transform
+
+        transformations = env.config.cautodoc_transformations
+        tropt = self.options.get('transform', OPT_DEFAULT)
+
+        # Local option to skip transformation.
+        if tropt == OPT_NONE:
+            return None
+
+        if transformations is None:
+            if tropt != OPT_DEFAULT:
+                self.logger.warning('transform specified without cautodoc_transformations config.',
+                                    location=(env.docname, self.lineno))
```

Comment:
If the user specifies `none` it won't trigger this check, although that could arguably be the intended behaviour. Still, overall, I think this function is not particularly clear, and it handles a couple of cases in ways that may be surprising:
- The fact that `default` defaults to "no transformation" is neither documented nor obvious when reading this function.
- If the user specifies a `none` option in `cautodoc_transformations`, that is not flagged as an error, but this code will ignore it all the same which could be really puzzling for the user. Still on this topic, `none` may (mere supposition) also carry different meanings in different languages as a word or acronym, so its absence should either be enforced or we need to consider it as a potential valid key.

Suggestion (totally untested):
```python
    def __get_transform(self):
        env = self.state.document.settings.env

        # Special case values for transform.
        OPT_NONE = 'none'
        OPT_DEFAULT = 'default'

        # Handle deprecated compat. To be removed.
        transform = self.__get_compat_transform()
        if transform is not None:
            return transform

        transformations = env.config.cautodoc_transformations
        tropt = self.options.get('transform', OPT_DEFAULT)

        if transformations is None and tropt != OPT_DEFAULT:
            self.logger.warning('transform specified without cautodoc_transformations config.',
                                location=(env.docname, self.lineno))
            return None

        # Use an identity transformation for the none and default
        # transformations by default.
        if transformations.get(OPT_NONE) is None:
            transformations[OPT_NONE] = lambda comment: comment
        if transformations.get(OPT_DEFAULT) is None:
            transformations[OPT_DEFAULT] = lambda comment: comment

        transform = transformations.get(tropt)
        if transform is None:
            self.logger.warning(f'unknown transformation "{tropt}".',
                                location=(env.docname, self.lineno))
            return None

        return lambda comment: transform(comment, transform=tropt)
```

Reply:
I know, I was struggling with this a bit, trying to make it least surprising.

It's intentional that `none` and `default` don't flag an error if `cautodoc_transformations` is not specified. They are the special cases.

The downside with adding `none` and `default` to transformation is that then all comments go through the no-op lambda, while it could be avoided by returning `None`.

Reply:
I'm trying to say, I'm not completely happy with my original function, I agree with your criticism, but I'm also not happy with your suggestion. Need to think about this.

Reply:
Okay, so here's a suggestion to remove `none` as a special case: allow `None` as a no-op "function" in `cautodoc_transformations`:

```
cautodoc_transformations = {
    'foo' = None,
}
```

Using `foo` in the directive transform option would explicitly do no transformation. This 1) skips the lambda when no transformation is needed, and 2) removes the special case key.

Reply:
Is it bad form to use `None` as a dict key for the default case? You wouldn't be able to explicitly choose the default case (because you can only have a string in the directive option), but I don't think that's a big deal.

I guess alternatively the empty string as a key could work be made to work too?

Reply:
This is what it would look like (the options part is not resolved yet). Unless using `None` as key is badly frowned on, I'd roll with this.

```python
    def __get_transform(self):
        env = self.state.document.settings.env

        # Handle deprecated compat. To be removed.
        transform = self.__get_compat_transform()
        if transform is not None:
            return transform

        transformations = env.config.cautodoc_transformations
        tropt = self.options.get('transform')

        if transformations is None:
            if tropt is not None:
                self.logger.warning('transform specified without cautodoc_transformations config.',
                                    location=(env.docname, self.lineno))

            return None

        # Note: None is a valid key for default.
        if tropt not in transformations:
            if tropt is not None:
                self.logger.warning(f'unknown transformation "{tropt}".',
                                    location=(env.docname, self.lineno))
            return None

        # Note: None is a valid value for no transformation.
        transform = transformations.get(tropt)
        if transform is None:
            return None

        return lambda comment: transform(comment, transform=tropt)

```

Reply:
#50 is what everything would look like with this change.

Reply:
@BrunoMSantos I hate to ping for feedback, especially as I haven't been exactly fast with my responses either, but do you think you'll be able to spare a moment to look at this? I'd really like to get this out the door and do a release.

Reply:
Obviously I could just push, but I value your feedback!

Reply:
Yeah, I know I'm falling behind again, real life gets in the way sometimes. I'll have a look today after work then.